src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs

```diff
@@ -2,150 +2,138 @@
             _launchProviders = launchProviders;
         }
 
         [ProjectAutoLoad(startAfter: ProjectLoadCheckpoint.ProjectFactoryCompleted)]
         [AppliesTo(ProjectCapability.CSharpOrVisualBasicOrFSharp)]
-        internal Task Load()
+        public Task InitializeAsync()
         {
-            EnsureInitialized();
-            return Task.CompletedTask;
+            return InitializeAsync(CancellationToken.None);
         }
 
-        protected override void Dispose(bool disposing)
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
         {
-            if (disposing)
-            {
-                _evaluationSubscriptionLink?.Dispose();
-            }
-        }
+            _projectGuid = await _projectGuidService.GetProjectGuidAsync()
+                                                    .ConfigureAwait(false);
 
-        protected override void Initialize()
-        {
-            var watchedEvaluationRules = Empty.OrdinalIgnoreCaseStringSet.Add(ConfigurationGeneral.SchemaName);
-            var evaluationBlock = new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(ConfigurationGeneralRuleBlock_ChangedAsync);
-            _evaluationSubscriptionLink = _activeConfiguredProjectSubscriptionService.ProjectRuleSource.SourceBlock
-                                            .LinkTo(target: evaluationBlock, ruleNames: watchedEvaluationRules, suppressVersionOnlyUpdates: true);
-        }
+            Assumes.False(_projectGuid == Guid.Empty);
 
-        internal async Task ConfigurationGeneralRuleBlock_ChangedAsync(IProjectVersionedValue<IProjectSubscriptionUpdate> e)
-        {
-            IProjectChangeDescription projectChange = e.Value.ProjectChanges[ConfigurationGeneral.SchemaName];
+            _startupProjectsListService = (IVsStartupProjectsListService)await _serviceProvider.GetServiceAsync(typeof(SVsStartupProjectsListService))
+                                                                                               .ConfigureAwait(false);
 
-            if (projectChange.Difference.ChangedProperties.Contains(ConfigurationGeneral.ProjectGuidProperty))
-            {
-                if (Guid.TryParse(projectChange.After.Properties[ConfigurationGeneral.ProjectGuidProperty], out Guid result))
-                {
-                    _guid = result;
-                    await AddOrRemoveProjectFromStartupProjectListAsync(initialize: true).ConfigureAwait(false);
-                }
+            Assumes.Present(_startupProjectsListService);
 
-                return;
-            }
+            _subscription = _projectSubscriptionService.ProjectRuleSource.SourceBlock.LinkTo(
+                target: new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(OnProjectChangedAsync),
+                suppressVersionOnlyUpdates: true,
+                linkOptions: new DataflowLinkOptions() { PropagateCompletion = true });
```

Comment:
Do you need to capture capabilities here?

Reply:
Good point. @lifengl thoughts? We read from a OrderPrecedenceImportCollection from the active configuration in the handler.

Reply:
I filed https://github.com/dotnet/project-system/issues/3322 to track this; I didn't change when we got initialized or when we walk the debuggers, so if the capabilities were wrong, they were always wrong. I'll do a pass over every dataflow usage and figure out if we need start a new context.